### ISSUE TITLE
Can type annotations be ignored?

### ISSUE THREAD

**Issue by michaelboulton** [facebook/reason#2405]
I'm not actually sure if this is expected behaviour but I ran into a weird issue where the type annotations were being ignored on one of my functions. Here is me mousing over it in vscode:

![reasonbug](https://user-images.githubusercontent.com/3809868/57218691-ddc33180-6fed-11e9-8a53-7ac42ffcdabe.png)

I have explicit type annotations on that function, but they seem to be being ignored.

This is normally fine, but then if I define a module like

```reason
module type Binder = {
  type t('a);
  let bind: (t('a), 'a => t('b)) => t('b);
};
```
And try to use it
```reason
  module PM: Binder = {
    type t('a) = property_t('a);
    let bind = bind;
  };
```
Then I get an error

```
This module doesn't match its signature because the value bind has the wrong type

  At library/Binder.re:22
  the signature required that bind be of type:

  (t('a), 'a => t('b)) => t('b)


  At library/Property.re:471
  your module defined bind having type:

  (property_t('a), 'a => property_t('a)) => property_t('a)

```

Is this expected behaviour that type annotations can be ignored? If so, is there some way to stop it doing this? If it's not I can try to create a small MWE.

**Comment by michaelboulton** [comment]
To be clear, I would expect it to give an error inside the `bind` function saying that the return value was the wrong type or something of that nature

**Comment by michaelboulton** [comment]
It turns out I can force a type error by using locally abstract types:

```reason
  let bind:
    type ta tb. (property_t(ta), ta => property_t(tb)) => property_t(tb) =
    (m, func) => {
...
```

This will complain that the return type from inside the function is wrong as expected. I'm still not sure why the type checker will allow one but not the other when they logically seem to be equal?

**Comment by anmonteiro** [comment]
Reason doesn't act at the type checking level, just purely at the syntax level, so this is in fact an OCaml question.

I'm interested in helping you but I can't answer your question without knowing where `property_t('a)` and `bind` come from.

**Comment by ubsan** [comment]
@michaelboulton unfortunately, they are not equivalent; `'a` is a type variable in let bindings, not a type parameter, which means that:

```
let x: 'a -> 'a = ...
```

means that `x` is of the form (for some `'a`), `'a -> 'a`. If OCaml can give it the type `type a. a -> a`, it will, but if you write

```ml
let x : 'a -> 'a = function
  | Some i -> Some (i + 1)
  | None -> None
```

OCaml will happily replace `'a` with `int option`.

**Comment by michaelboulton** [comment]
@anmonteiro I suppose I worded it badly in the original post, but I was also wondering whether this might have been some difference between Reason/OCaml which it doesn't seem to be. I saw in my code after adding the explicit type annotation where the error was, I was just confused that it was happening at all.

I assumed that the type checker would not convert a polymorphic type to another polymorphic type (I assumed it would only ever convert from something like `'a` to `int`, not `'a` to `'b`, if that makes sense) but I can see from @ubsan 's answer that it can in fact do this.

Thanks for the help, I'll close this now.
